File: ios/Gogai/Models/FieldUpdate.swift
import Foundation

/// nil への更新と「変更しない」を型安全に区別するための汎用 enum
enum FieldUpdate<T: Sendable>: Sendable {
    case keep
    case clear
    case set(T)

    func apply(to current: T?) -> T? {
        switch self {
        case .keep: return current
        case .clear: return nil
        case .set(let value): return value
        }
    }
}

File: android/app/src/main/kotlin/com/mtkg/gogai/model/FieldUpdate.kt
package com.mtkg.gogai.model

/// nil への更新と「変更しない」を型安全に区別するための汎用 sealed class（iOS FieldUpdate の移植）
sealed class FieldUpdate<out T> {
    data object Keep : FieldUpdate<Nothing>()
    data object Clear : FieldUpdate<Nothing>()
    data class Set<T>(val value: T) : FieldUpdate<T>()

    fun applyTo(current: @UnsafeVariance T?): T? = when (this) {
        is Keep -> current
        is Clear -> null
        is Set -> value
    }
}
